Property [autoIncrement] was not found for object type [liquibase.datatype.core.BigIntType]

### Describe the bug

Liquibase crashed at runtime after the upgrade from 3.13.3 to 3.14.1.

### Expected behavior

Libquibase changes should applied normally.

### Actual behavior

Changeset fragment:

```xml
            <column name="id" type="bigint" autoIncrement="true">
                <constraints primaryKey="true" nullable="false"/>
            </column>
```

Stacktrace:

```log
 [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector] 2024-08-31 09:31:28,050 ERROR [io.qua.run.Application] (main) Failed to start application: java.lang.RuntimeException: Failed to start quarkus
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at io.quarkus.runtime.Application.start(Application.java:101)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:119)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at io.quarkus.runtime.Quarkus.run(Quarkus.java:44)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at io.quarkus.runtime.Quarkus.run(Quarkus.java:124)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at io.quarkus.runner.GeneratedMain.main(Unknown Source)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector] Caused by: java.lang.IllegalStateException: Error starting Liquibase
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at io.quarkus.liquibase.runtime.LiquibaseRecorder.doStartActions(LiquibaseRecorder.java:93)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at io.quarkus.deployment.steps.LiquibaseProcessor$startLiquibase1744275855.deploy_0(Unknown Source)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at io.quarkus.deployment.steps.LiquibaseProcessor$startLiquibase1744275855.deploy(Unknown Source)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  ... 7 more
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector] Caused by: liquibase.exception.CommandExecutionException: liquibase.exception.LiquibaseException: liquibase.exception.UnexpectedLiquibaseException: Property [autoIncrement] was not found for object type [liquibase.datatype.core.BigIntType]
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.command.CommandScope.execute(CommandScope.java:258)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.Liquibase.lambda$validate$24(Liquibase.java:1247)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.Scope.lambda$child$0(Scope.java:191)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.Scope.child(Scope.java:200)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.Scope.child(Scope.java:190)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.Scope.child(Scope.java:169)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.Liquibase.runInScope(Liquibase.java:1290)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.Liquibase.validate(Liquibase.java:1242)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at io.quarkus.liquibase.runtime.LiquibaseRecorder.doStartActions(LiquibaseRecorder.java:82)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  ... 9 more
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector] Caused by: liquibase.exception.LiquibaseException: liquibase.exception.UnexpectedLiquibaseException: Property [autoIncrement] was not found for object type [liquibase.datatype.core.BigIntType]
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:148)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.changelog.DatabaseChangeLog.validate(DatabaseChangeLog.java:381)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.command.core.helpers.DatabaseChangelogCommandStep.run(DatabaseChangelogCommandStep.java:92)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.command.CommandScope.execute(CommandScope.java:220)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  ... 17 more
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector] Caused by: liquibase.exception.UnexpectedLiquibaseException: Property [autoIncrement] was not found for object type [liquibase.datatype.core.BigIntType]
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.util.ObjectUtil.setProperty(ObjectUtil.java:128)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.datatype.DataTypeFactory.fromDescription(DataTypeFactory.java:251)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.change.core.CreateTableChange.generateStatements(CreateTableChange.java:101)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.change.AbstractChange.generateStatementsVolatile(AbstractChange.java:352)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector] Exception in thread "Shutdown thread" java.lang.NullPointerException
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.change.AbstractChange.warn(AbstractChange.java:423)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.changelog.visitor.ValidatingVisitor.validateChange(ValidatingVisitor.java:159)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.changelog.visitor.ValidatingVisitor.visit(ValidatingVisitor.java:119)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at io.quarkus.runtime.ApplicationLifecycleManager$ShutdownHookThread.run(ApplicationLifecycleManager.java:455)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.changelog.ChangeLogIterator.lambda$run$0(ChangeLogIterator.java:131)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:896)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:872)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.Scope.lambda$child$0(Scope.java:191)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.Scope.child(Scope.java:200)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.Scope.child(Scope.java:190)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.Scope.child(Scope.java:169)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.changelog.ChangeLogIterator.lambda$run$1(ChangeLogIterator.java:120)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.Scope.lambda$child$0(Scope.java:191)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.Scope.child(Scope.java:200)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.Scope.child(Scope.java:190)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.Scope.child(Scope.java:169)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.Scope.child(Scope.java:257)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.Scope.child(Scope.java:261)
      > [cdr-log-collector-7b85bc6fc4-c9kxz cdr-log-collector]  at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:89)

```

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

K8s 1.30.1

### Output of `java -version`

21.0.4+7-LTS

### Mandrel or GraalVM version (if different from Java)

Mandrel-23.1.4.0-Final

### Quarkus version or git rev

3.14.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 8.10

### Additional information

Workaround:

```
@RegisterForReflection(targets = [BigIntType::class])
```

--- a/extensions/liquibase/deployment/src/main/java/io/quarkus/liquibase/deployment/LiquibaseProcessor.java
+++ b/extensions/liquibase/deployment/src/main/java/io/quarkus/liquibase/deployment/LiquibaseProcessor.java
@@ -1,7 +1,5 @@
 package io.quarkus.liquibase.deployment;
 
-import static java.util.function.Predicate.not;
-
 import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.nio.file.Paths;
@@ -16,7 +14,6 @@ import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.BiConsumer;
-import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
 import jakarta.enterprise.context.ApplicationScoped;
@@ -36,7 +33,6 @@ import io.quarkus.arc.deployment.BeanContainerBuildItem;
 import io.quarkus.arc.deployment.SyntheticBeanBuildItem;
 import io.quarkus.arc.processor.DotNames;
 import io.quarkus.datasource.common.runtime.DataSourceUtil;
-import io.quarkus.deployment.Capabilities;
 import io.quarkus.deployment.Feature;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
@@ -77,6 +73,7 @@ import liquibase.change.core.SQLFileChange;
 import liquibase.changelog.ChangeLogParameters;
 import liquibase.changelog.ChangeSet;
 import liquibase.changelog.DatabaseChangeLog;
+import liquibase.datatype.DataTypeInfo;
 import liquibase.exception.LiquibaseException;
 import liquibase.parser.ChangeLogParser;
 import liquibase.parser.ChangeLogParserFactory;
@@ -97,6 +94,7 @@ class LiquibaseProcessor {
             "www.liquibase.org/xml/ns/dbchangelog/*.xsd"));
 
     private static final DotName DATABASE_CHANGE_PROPERTY = DotName.createSimple(DatabaseChangeProperty.class.getName());
+    private static final DotName DATA_TYPE_INFO_ANNOTATION = DotName.createSimple(DataTypeInfo.class.getName());
 
     @BuildStep
     FeatureBuildItem feature() {
@@ -114,7 +112,6 @@ class LiquibaseProcessor {
             List<JdbcDataSourceBuildItem> jdbcDataSourceBuildItems,
             CombinedIndexBuildItem combinedIndex,
             CurateOutcomeBuildItem curateOutcome,
-            Capabilities capabilities,
             BuildProducer<ReflectiveClassBuildItem> reflective,
             BuildProducer<NativeImageResourceBuildItem> resource,
             BuildProducer<ServiceProviderBuildItem> services,
@@ -126,8 +123,19 @@ class LiquibaseProcessor {
                 liquibase.sqlgenerator.core.LockDatabaseChangeLogGenerator.class.getName()));
 
         reflective.produce(ReflectiveClassBuildItem
-                .builder(liquibase.change.AbstractSQLChange.class, liquibase.database.jvm.JdbcConnection.class).methods()
+                .builder(liquibase.datatype.core.UnknownType.class.getName())
                 .reason(getClass().getName())
+                .constructors().methods()
+                .build());
+
+        reflective.produce(ReflectiveClassBuildItem.builder(
+                liquibase.AbstractExtensibleObject.class.getName(),
+                liquibase.change.core.DeleteDataChange.class.getName(),
+                liquibase.change.core.EmptyChange.class.getName(),
+                liquibase.database.jvm.JdbcConnection.class.getName(),
+                liquibase.plugin.AbstractPlugin.class.getName())
+                .reason(getClass().getName())
+                .methods()
                 .build());
 
         reflective.produce(ReflectiveClassBuildItem
@@ -146,6 +154,7 @@ class LiquibaseProcessor {
                 .constructors().build());
 
         reflective.produce(ReflectiveClassBuildItem.builder(
+                liquibase.changelog.RanChangeSet.class.getName(),
                 liquibase.configuration.LiquibaseConfiguration.class.getName(),
                 liquibase.parser.ChangeLogParserConfiguration.class.getName(),
                 liquibase.GlobalConfiguration.class.getName(),
@@ -185,6 +194,18 @@ class LiquibaseProcessor {
                         .reason(getClass().getName())
                         .constructors().methods().fields().build());
 
+        // register all liquibase.datatype.core.* data types
+        Set<String> classesAnnotatedWithDataTypeInfo = combinedIndex.getIndex().getAnnotations(DATA_TYPE_INFO_ANNOTATION)
+                .stream()
+                .map(AnnotationInstance::target)
+                .filter(at -> at.kind() == AnnotationTarget.Kind.CLASS)
+                .map(at -> at.asClass().name().toString())
+                .collect(Collectors.toSet());
+        reflective.produce(ReflectiveClassBuildItem.builder(classesAnnotatedWithDataTypeInfo.toArray(String[]::new))
+                .reason(getClass().getName())
+                .constructors().methods()
+                .build());
+
         Collection<String> dataSourceNames = jdbcDataSourceBuildItems.stream()
                 .map(JdbcDataSourceBuildItem::getName)
                 .collect(Collectors.toSet());
@@ -200,20 +221,6 @@ class LiquibaseProcessor {
                     .constructors().methods().fields().build());
         });
 
-        // CommandStep implementations are needed
-        consumeService(liquibase.command.CommandStep.class.getName(), (serviceClassName, implementations) -> {
-            var filteredImpls = implementations.stream()
-                    .filter(commandStepPredicate(capabilities))
-                    .toArray(String[]::new);
-            services.produce(new ServiceProviderBuildItem(serviceClassName, filteredImpls));
-            reflective.produce(ReflectiveClassBuildItem.builder(filteredImpls)
-                    .reason(getClass().getName())
-                    .constructors().build());
-            for (String implementation : filteredImpls) {
-                runtimeInitialized.produce(new RuntimeInitializedClassBuildItem(implementation));
-            }
-        });
-
         var dependencies = curateOutcome.getApplicationModel().getRuntimeDependencies();
         resource.produce(NativeImageResourceBuildItem.ofDependencyResources(
                 dependencies, LIQUIBASE_ARTIFACT, LIQUIBASE_RESOURCE_FILTER));
@@ -221,14 +228,6 @@ class LiquibaseProcessor {
 
         // liquibase resource bundles
         resourceBundle.produce(new NativeImageResourceBundleBuildItem("liquibase/i18n/liquibase-core"));
-    }
-
-    private static Predicate<String> commandStepPredicate(Capabilities capabilities) {
-        if (capabilities.isPresent("io.quarkus.jdbc.h2")) {
-            return (s) -> true;
-        } else {
-            return not("liquibase.command.core.StartH2CommandStep"::equals);
-        }
     }
 
     private void consumeService(String serviceClassName, BiConsumer<String, Collection<String>> consumer) {

--- a/integration-tests/liquibase/src/main/resources/db/xml/changeLog.xml
+++ b/integration-tests/liquibase/src/main/resources/db/xml/changeLog.xml
@@ -5,9 +5,8 @@
                         https://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-latest.xsd">
 
     <include relativeToChangelogFile="true" file="create-tables.xml" />
-
     <include relativeToChangelogFile="true" file="test/test.xml" />
-
+    <include relativeToChangelogFile="true" file="test/test-data-types.xml" />
     <include relativeToChangelogFile="true" file="create-views.xml" />
 
     <includeAll relativeToChangelogFile="true" path="includeAll" />

--- a/integration-tests/liquibase/src/main/resources/db/xml/test/test-data-types.xml
+++ b/integration-tests/liquibase/src/main/resources/db/xml/test/test-data-types.xml
@@ -1,0 +1,21 @@
+<?xml version="1.1" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+                                       https://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-latest.xsd">
+
+    <changeSet author="dev (generated)" id="create-data-types-tables-1">
+        <createTable tableName="TYPE_TEST">
+            <column name="id" type="bigint" autoIncrement="true">
+                <constraints primaryKey="true" nullable="false"/>
+            </column>
+            <column name="decimal_col" type="decimal" autoIncrement="true"/>
+            <column name="int_col" type="int" autoIncrement="true"/>
+            <column name="med_int_col" type="mediumint" autoIncrement="true"/>
+            <column name="number_col" type="number" autoIncrement="true"/>
+            <column name="small_int_col" type="smallint" autoIncrement="true"/>
+            <column name="tiny_int_col" type="tinyint" autoIncrement="true"/>
+        </createTable>
+    </changeSet>
+
+</databaseChangeLog>

--- a/integration-tests/liquibase/src/test/java/io/quarkus/it/liquibase/LiquibaseFunctionalityTest.java
+++ b/integration-tests/liquibase/src/test/java/io/quarkus/it/liquibase/LiquibaseFunctionalityTest.java
@@ -30,7 +30,7 @@ public class LiquibaseFunctionalityTest {
                 .get("/liquibase/update")
                 .then()
                 .body(is(
-                        "create-tables-1,test-1,create-view-inline,create-view-file-abs,create-view-file-rel,"
+                        "create-tables-1,test-1,create-data-types-tables-1,create-view-inline,create-view-file-abs,create-view-file-rel,"
                                 + ("includeAll-1,includeAll-2,")
                                 + "json-create-tables-1,json-test-1,"
                                 + "sql-create-tables-1,sql-test-1,"

